### PR TITLE
SerializableErrorErrorDecoder wraps remote exception in local excepti…

### PR DESCRIPTION
**Pull request by uschi2000**
…on in order to produce local stacktrace

@pnepywoda @markelliot

### CODE REVIEW

File: http-clients/src/main/java/com/palantir/remoting/http/errors/SerializableErrorErrorDecoder.java [hunk 1]
```diff
@@ -59,12 +60,27 @@ public Exception decode(String methodKey, Response response) {
...
+
+                    // Construct local exception that wraps the remote exception
+                    Constructor<? extends Exception> constructor;
+                    try {
+                        constructor = exceptionClass.getConstructor(String.class, Throwable.class);
+                    } catch (NoSuchMethodException e) {
+                        constructor = exceptionClass.getConstructor(String.class);
                     }
-                    return exception;
+                    Exception localException = constructor.newInstance(error.getMessage(), remoteException);
```
Comment (pnepywoda) [inline http-clients/src/main/java/com/palantir/remoting/http/errors/SerializableErrorErrorDecoder.java]: doesn't this break on exception classes that only have String and not (String, Throwable) ?

Comment (uschi2000) [inline http-clients/src/main/java/com/palantir/remoting/http/errors/SerializableErrorErrorDecoder.java]: ah shoot, i thought of that case and then implemented it wrong ... (see try/catch above)

Comment (uschi2000) [inline http-clients/src/main/java/com/palantir/remoting/http/errors/SerializableErrorErrorDecoder.java]: try again

**Comment by pnepywoda** [comment]
Seems good. Do we care about exceptions that only have default constructors?

File: http-clients/src/main/java/com/palantir/remoting/http/errors/SerializableErrorErrorDecoder.java [hunk 2]
```diff
@@ -59,12 +59,27 @@ public Exception decode(String methodKey, Response response) {
                 try {
                     SerializableError error = MAPPER.readValue(bodyAsString, SerializableError.class);
                     Class<? extends Exception> exceptionClass = error.getExceptionClass();
-                    Exception exception = exceptionClass.getConstructor(String.class).newInstance(error.getMessage());
+
+                    // Construct remote exception
+                    Exception remoteException =
+                            exceptionClass.getConstructor(String.class).newInstance(error.getMessage());
```
Comment (markelliot) [inline http-clients/src/main/java/com/palantir/remoting/http/errors/SerializableErrorErrorDecoder.java]: we probably ought to do this construction and the one on line 74 in a utility method that safely tries to construct each of these and failing that just produces a RuntimeException holding data and shoves the classname of the uninstantiable exception into the error message.

**Comment by uschi2000** [comment]
@pnepywoda

**Pull request merged**